### PR TITLE
feat: add ProcessInstanceDiagramController Rest Api

### PR DESCRIPTION
**Pull request by igdianov**
Part of https://github.com/Activiti/Activiti/issues/2750

- [x] - feat: add canRead process definition key security policy check
- [x] - feat: add ProcessInstanceDiagramAdminController implementation endpoint 
`/admin/v1/process-instances/{processInstanceId}/diagram`
- [x] - feat: add ProcessInstanceDiagramController implementation endpoint 
`/v1/process-instances/{processInstanceId}/diagram`

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -43,6 +43,7 @@
 
   <properties>
     <activiti-cloud-build.version>7.1.7</activiti-cloud-build.version>
+    <activiti-dependencies.version>7.1.22</activiti-dependencies.version>
```
Comment (salaboy) [inline pom.xml]: @igdianov I sincerely think that this is a bad idea.. we are once again linking activiti-dependencies to query.. we need to make sure that no dependency from the engine is dragged here.. and we should rely only on APIs. I don't see the need of generating the SVG in the backend

**Review by salaboy — COMMENTED** [review]
@igdianov I am worried about two main things:
1) Separation of concerns and dependencies, query should return data, not process the generation of an SVG. Also the dependency comment that I've added below. 
2) Too many new entities (I would avoid having entities.. this is going to increate the amount of tables and relationships that we handle).. I think that we can do better by summarizing the data in a json structure, making sure that something like BPMN.io rendered widget had enough information to render and decorate the model plus the active and already executed activities. 

But this is just my opinion

File: pom.xml (hunk 1 above)
Comment (igdianov) [reply pom.xml]: @salaboy Thank you for feedback. To be able render any BPMN diagram we need to aggregate events for all BPMN activities in order to keep track of their status, i.e STARTED, COMPLETED, CANCELLED as well as sequence flows taken between them. Therefore, I have added two entities with corresponding event handlers for `BPMNActivityEntity` and `BPMNSequenceFlowEntity` to accommodate both rendering diagrams on the server with SVG now and client side with BPMN.io JS later. 

I don't think that aggregating activities into Json blob will be flexible enough to support both server and client side rendering as well as other query use cases. The BPMNActivity and SequenceFlow entities provide information about activity type, i.e. timer, signal, sub-process, start, stop, serviceTask. This allows to query and filter them by attributes such as status for any running or historical process instance. We can also use the structured data model to generate Json blob to accommodate client side rendering with BPMN.io and use GraphQL Query with Notifications to provide that functionality out-of-the-box right now.

I would also argue that query returning SVG is a data derived from historical record of the process instance state in the Query database.  This is makes it easy for anyone to simply embed SVG diagram into any web application via single Query Rest Api call. I would also defer to @magemello and Claire in order to articulate the need to support SVG rendering for enterprise customers and product. 

I agree that having explicit dependencies on the Activiti engine is not a good idea. The `activiti-dependencies` are only used to import managed dependencies required to parse, layout, and generate BPMN diagram only:

```
     <dependency>
      <groupId>org.activiti</groupId>
      <artifactId>activiti-image-generator</artifactId>
    </dependency>
    <dependency>
      <groupId>org.activiti</groupId>
      <artifactId>activiti-bpmn-converter</artifactId>
    </dependency>
    <dependency>
      <groupId>org.activiti</groupId>
      <artifactId>activiti-bpmn-layout</artifactId>
    </dependency>
```
Perhaps, these dependencies can be moved to https://github.com/Activiti/activiti-core-common for better separation of concerns.

Comment (Claire) [reply pom.xml]: wrong Claire mentioned

Comment (salaboy) [reply pom.xml]: @igdianov yup that was my understanding for looking at the code. 
I think that that for client side rendering storing the right information in json format is enough. 
I am 100% sure that clients will have access to JS libraries to render the SVG in the browser, or we can even provide an SVG file when the process gets deployed, and then just decorate with the information stored in JSON on the client side. 

This is exactly my point: **I would also defer to @magemello and Claire in order to articulate the need to support SVG rendering for enterprise customers and product.**, what is the use case, where we must support server side rendering? If we add these in the query backend then we are stuck providing that feature which will definitely impact the service responsibilities. The fact that we are not prioritizing the **BPMN.io JS** approach highlight the lack of understanding of impact that doing server side rendering of SVGs might have. Also notice this issue: https://github.com/Activiti/Activiti/issues/1602 which is being dragged forward by adding the modules that you mention, adding one more red flag to move forward with that approach.  

Regarding dependencies, the same.. I like your suggestion, but let's make sure that even if we point to a module under activiti.dependencies, we just point to the module, so I would start by moving those modules out of Activiti Core (github.com/activiti/activiti) and then pointing to the sub dependency.. in this case activiti-core-common.dependencies. Should we create an issue for moving those modules out of core?

**Review by erdemedeiros — APPROVED** [review]

**Pull request merged**